### PR TITLE
CI with NSX improvements

### PR DESCRIPTION
**Pull request by remibergsma**
NSX improvements for our CI setup:
- add support for XenServer next to KVM
- find the right master controller and use that everywhere
- fix hardcoded NSX controller ip that would make zone2 setup fail
- split into two configs: minimal with single nodes and full blown cluster config (both KVM and XenServer)
- resolved instability, now builds reliably where it would fail 1:7 or so times

WIP: Still testing all the different options but starts to look good!

### CODE REVIEW

**Comment by remibergsma** [comment]
This needs the latest version of Bubble Cookbook because the new hosts are added here: https://github.com/MissionCriticalCloud/bubble-cookbook/commit/057217b9acf1401adbee48fe8f0d2b582399a90a

File: deploy/default/role/xenserver62.conf [hunk 1]
```diff
@@ -2,10 +2,10 @@
...
-disk_bus = ide
+hardware = openvswitch
+net_model = rtl8139
+disk_dev = vda
+disk_bus = sata
 image = xen62
-firstboot = 
-postboot = 
+firstboot =
+postboot =
```
Comment (borisroman) [inline deploy/default/role/xenserver62.conf]: Newline

File: deploy/kvm_local_deploy.py [hunk 2]
```diff
@@ -563,8 +566,9 @@ def sort_nsx_vm_name(self, vm_name):
             return 4
 
     def deploy_cloud_roles(self, roles):
-        pool = ThreadPool(4)
+        pool = ThreadPool(10)
```
Comment (borisroman) [inline deploy/kvm_local_deploy.py]: Virtual bubbles don't like this.

File: deploy/kvm_local_deploy.py [hunk 3]
```diff
@@ -652,8 +657,9 @@ def delete_marvin(self):
             return False
         print "Note: Found hypervisor type '" + self.get_hypervisor_type() + "'"
         hosts = self.get_hosts()
-        pool = ThreadPool(4)
+        pool = ThreadPool(10)
```
Comment (borisroman) [inline deploy/kvm_local_deploy.py]: Virtual bubbles don't like this.

File: ci/ci-setup-infra.sh [hunk 4]
```diff
@@ -252,22 +251,22 @@ function create_nsx_cluster {
 function setup_nsx_cosmic {
 
   say "Generating script for setting up NSX controller in Cosmic"
-  echo "next_host_id=\$(mysql -h ${csip} -u cloud -pcloud cloud -e \"SELECT MAX(id) +1 FROM host;\" -s)" > /tmp/nsx_cosmic.sh
+  echo "set -x" > /tmp/nsx_cosmic.sh
```
Comment (borisroman) [inline ci/ci-setup-infra.sh]: `#!/usr/bin/env bash` 👍

**Review by borisroman — CHANGES_REQUESTED** [review]
Starts to look good! Few minor notes.

**Comment by remibergsma** [comment]
XenServer tests:

image

image

image

image

File: deploy/default/role/xenserver62.conf (hunk 1 above)
Comment (remibergsma) [reply deploy/default/role/xenserver62.conf]: done

**Comment by remibergsma** [comment]
KVM tests:

image

image

The other tests are still running...

**Comment by remibergsma** [comment]
Rebased with latest master

**Comment by borisroman** [comment]
@remibergsma Also remove this: https://github.com/MissionCriticalCloud/bubble-toolkit/blob/master/deploy/default/firstboot/configure_ovs.sh#L97

**Comment by remibergsma** [comment]
The issue where no master could be found and the build would hang seems resolved with the retry:

Initially:
```
++ curl -L -sD - -k -b /tmp/nsx_cookie_14900.txt https://nsxcon4/ws.v1/control-cluster
++ egrep 'HTTP/1.1 200'
+ is_master=
+ '[' 1 -gt 0 ']'
+ say 'Controller nsxcon4 DOES NOT respond with 200 so this is NOT our master!'
+ echo '==> Controller nsxcon4 DOES NOT respond with 200 so this is NOT our master!'
==> Controller nsxcon4 DOES NOT respond with 200 so this is NOT our master!
+ say 'Output: '
+ echo '==> Output: '
==> Output:
```

In retry:

```
+ say 'Master not yet found, sleeping 10 sec and trying again..'
+ echo '==> Master not yet found, sleeping 10 sec and trying again..'
==> Master not yet found, sleeping 10 sec and trying again..
+ sleep 10
+ :
+ for i in 1 2 3 4 5 6 7 8 9
++ eval 'echo ${nsx_controller_node_ip1}'
+++ echo nsxcon4
+ '[' '!' -v nsxcon4 ']'
+ nsx_controller_node_ip=
+ eval 'nsx_controller_node_ip=${nsx_controller_node_ip1}'
++ nsx_controller_node_ip=nsxcon4
+ say 'Checking to see if nsxcon4 is master'
+ echo '==> Checking to see if nsxcon4 is master'
==> Checking to see if nsxcon4 is master
+ say 'Authenticating against NSX controller nsxcon4'
+ echo '==> Authenticating against NSX controller nsxcon4'
==> Authenticating against NSX controller nsxcon4
+ curl -L -k -c /tmp/nsx_cookie_14900.txt -X POST -d 'username=admin&password=admin' https://nsxcon4/ws.v1/login
Successful Authentication.

You successfully authenticated.  Use the cookie in this reply in future requests.
++ curl -L -sD - -k -b /tmp/nsx_cookie_14900.txt https://nsxcon4/ws.v1/control-cluster
++ egrep 'HTTP/1.1 200'
' is_master='HTTP/1.1 200 OK
+ '[' 0 -gt 0 ']'
+ say 'Controller nsxcon4 responds with 200 so this is our master!'
+ echo '==> Controller nsxcon4 responds with 200 so this is our master!'
==> Controller nsxcon4 responds with 200 so this is our master!
```

**Comment by remibergsma** [comment]
The 2nd KVM setup works fine except for the nsxsvc4. Still investigating why that happens.

image

**Comment by remibergsma** [comment]
It's something with the name, because with nsxsvc[23] it works fine ;-) 
image

**Review by borisroman — APPROVED** [review]

**Pull request merged**